### PR TITLE
Ficus hf variant

### PR DESCRIPTION
**Pull request by JakubFornadel**
## Purpose

Changes:
- Pillar blocks are created each N pbft blocks
- Pillar block contain state root & bridge epoch from period N-5
- Pillar votes for pillar block N have period N+1
- pbft block N+1 is not finalized unless we have also > 50% of pillar votes with period N+1

This change fixes:
- potentially getting stalled due to missing pillar chain votes that would no longer be generated
- asserts are not triggered anymore
- the latest set of validators control pillar chain consensus

This change requires:
- testnet reset
- TaraClient contract change: https://github.com/Taraxa-project/bridge/blob/master/src/eth/TaraClient.sol#L105 - `blocks[i].block.period + 1` should be used

evm pr https://github.com/Taraxa-project/taraxa-evm/pull/242

### CODE REVIEW

File: libraries/core_libs/consensus/include/pbft/pbft_manager.hpp [hunk 1]
```diff
@@ -598,6 +598,7 @@ class PbftManager {
   bool already_next_voted_null_block_hash_ = false;
   bool go_finish_state_ = false;
   bool loop_back_finish_state_ = false;
+  PbftPeriod last_placed_pillar_vote_period_ = 0;
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/pbft/pbft_manager.hpp:601]: yes

File: libraries/core_libs/consensus/src/final_chain/final_chai_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,588 @@
+#include "common/encoding_solidity.hpp"
+#include "final_chain/final_chain_impl.hpp"
+#include "final_chain/trie_common.hpp"
+
+namespace taraxa::final_chain {
+
+FinalChainImpl::FinalChainImpl(const std::shared_ptr<DB>& db, const taraxa::FullNodeConfig& config,
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/final_chain/final_chai_impl.cpp:7]: nothing new, just moved implementation from header file, virtual is there from the beginning - we can remove virtual in separate task

File: libraries/config/src/hardfork.cpp [hunk 3]
```diff
@@ -73,43 +74,30 @@ void FicusHardforkConfig::validate(uint32_t delegation_delay) const {
...
+    throw taraxa::ConfigException("ficus_hf.pillar_blocks_interval must be > 1");
   }
 
-  if (pillar_chain_sync_interval <= pbft_inclusion_delay) {
-    throw taraxa::ConfigException("ficus_hf.pillar_chain_sync_interval must be > ficus_hf.pbft_inclusion_delay");
-  }
-
-  if (pbft_inclusion_delay < 1 || pbft_inclusion_delay <= delegation_delay) {
-    throw taraxa::ConfigException("ficus_hf.pbft_inclusion_delay must be >= 1 && > dpos.delegation_delay");
+  if (block_num % pillar_blocks_interval) {
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/config/src/hardfork.cpp:81]: I don't remember. There was something in code that required it... Anyway it does not hurt anything except making sure the config value are correct, which happens only once

File: libraries/core_libs/consensus/src/final_chain/final_chai_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+                  .code_retval)
+      .convert_to<bool>();
+}
+
+std::vector<SharedTransaction> FinalChainImpl::makeSystemTransactions(PbftPeriod blk_num) {
+  std::vector<SharedTransaction> system_transactions;
+  // Make system transactions <delegation_delay()> blocks sooner than next pillar block period,
+  // e.g.: if pillar block period is 100, this will return true for period 100 - delegation_delay() == 95, 195, 295,
+  // etc...
+  if (kHardforksConfig.ficus_hf.isPillarBlockPeriod(blk_num + delegation_delay())) {
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/final_chain/final_chai_impl.cpp:134]: @kstdl it this change ok ? Doesn't t break something in contract ?

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 5]
```diff
@@ -1764,17 +1784,58 @@ bool PbftManager::pushPbftBlock_(PeriodData &&period_data, std::vector<std::shar
...
+  // Pillar block use state from current_pbft_chain_size - final_chain_->delegation_delay(), e.g. block with period 32
+  // uses state from period 27.
+  PbftPeriod request_period = current_pbft_chain_size - final_chain_->delegation_delay();
+  // advancePeriod() -> resetConsensus() -> waitForPeriodFinalization() makes sure block request_period was already
+  // finalized
+  assert(final_chain_->last_block_number() >= request_period);
+
+  const auto block_header = final_chain_->block_header(request_period);
+  const auto bridge_root = final_chain_->get_bridge_root(request_period);
+  const auto bridge_epoch = final_chain_->get_bridge_epoch(request_period);
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1807]: @kstdl it this change ok ? Doesn't t break something in contract ? Check `PbftPeriod request_period = current_pbft_chain_size - final_chain_->delegation_delay();`

**Review by mfrankovi — APPROVED** [review]

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 5 above)
Comment (kstdl) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1813]: yeah. It should be just after that system transaction which is a case here if I am not mistaken

**Pull request merged**